src/solvers/td3.ml

```diff
@@ -113,49 +104,50 @@ module WP =
             HM.mem called y || destabilize_vs y || b || was_stable && List.mem y vs
           ) w false
       and solve ?reuse_eq x phase =
-        if tracing then trace "sol2" "solve %a, called: %b, stable: %b\n" S.Var.pretty_trace x (HM.mem called x) (HM.mem stable x);
+        if tracing then trace "sol2" "solve %a, called: %b, stable: %b, wpoint: %b\n" S.Var.pretty_trace x (HM.mem called x) (HM.mem stable x) (HM.mem wpoint x);
         init x;
         assert (S.system x <> None);
         if not (HM.mem called x || HM.mem stable x) then (
           HM.replace stable x ();
           HM.replace called x ();
-          let wp = HM.mem wpoint x in
-          let old = HM.find rho x in
-          let l = HM.create 10 in
-          let tmp =
+          let l = HM.create 10 in (* local cache *)
+          let wp = HM.mem wpoint x in (* if x becomes a wpoint during eq, checking this will delay widening until next solve *)
+          let old = HM.find rho x in (* d from older solve *)
+          let eqd = (* d from equation/rhs *)
             match reuse_eq with
             | Some d -> d
             | None -> eq x (eval l x) (side ~x)
           in
-          let new_eq = tmp in
-          (* let tmp = if GobConfig.get_bool "ana.opt.hashcons" then S.Dom.join (S.Dom.bot ()) tmp else tmp in (* Call hashcons via dummy join so that the tag of the rhs value is up to date. Otherwise we might get the same value as old, but still with a different tag (because no lattice operation was called after a change), and since Printable.HConsed.equal just looks at the tag, we would uneccessarily destabilize below. Seems like this does not happen. *) *)
-          if tracing then trace "sol" "Var: %a\n" S.Var.pretty_trace x ;
-          if tracing then trace "sol" "Contrib:%a\n" S.Dom.pretty tmp;
           HM.remove called x;
-          let tmp =
-            if not wp then tmp
+          let wpd = (* d after widen/narrow (if wp) *)
+            if not wp then eqd
             else
               if term then
-                match phase with Widen -> S.Dom.widen old (S.Dom.join old tmp) | Narrow -> S.Dom.narrow old tmp
+                match phase with
+                | Widen -> S.Dom.widen old (S.Dom.join old eqd)
+                | Narrow ->
+                  assert S.Dom.(leq eqd old || not (leq old eqd)); (* eqd > old would be unsound, but can be incomparable *)
```

Comment:
When can it be incomparable?

Also, is this assertion with two extra `leq` checks something we want to keep here? Was there some issue with this at some point?

Reply:
On sv-benchmarks this assertion fails on 7 benchmarks:
![image](https://user-images.githubusercontent.com/378740/206282650-594d56e9-1a40-4bae-9a4c-d0638161faeb.png)
There's also a ~1% slowdown from this extra checking.

I will try looking into these failures to see what's going on there.

Reply:
It's hard to trace the solving of so large programs, but at least at the point of this assert failing there appear to be some odd values. In the narrowing phase, somehow intdomains, address sets, partitioned arrays and var_eq equivalences can suddenly increase. This happens even with the witness lifting disabled.

I think we can keep the assertion because it is still pointing out some potential issues. Such increases in the narrowing phase might also be the reason for their non-termination and are thus worth looking into separately.

This might somehow be related to the fact that `equal` isn't always equivalent to two-way `leq`. We use `equal` to detect the end of the widening phase and switch to narrowing, after which the `leq` in this assertion confusingly no longer holds.

Reply:
We decided in Gobcon to remove this assert.

Reply:
I will remove the assertion for merge. It was discussed at GobCon that in non-monotonic systems this isn't necessarily an issue, but just something which theoretically can happen and hasn't been previously observed.